### PR TITLE
HDDS-8134. Certificate clients are not correctly closed.

### PR DESCRIPTION
**Pull request by fapifta**
## What changes were proposed in this pull request?

[HDDS-7874](https://issues.apache.org/jira/browse/HDDS-7874) revealed that the Certificate client instances in the system are not closed properly.
This PR adds the close() call for all CertificateClient instances that are created either in tests or during normal operations, to ensure that the internal certificate lifetime monitor task executor thread is shut down correctly once the client is not used anymore.

## What is the link to the Apache JIRA

https://issues.apache.org/jira/browse/HDDS-8134

## How was this patch tested?
Existing tests should cover.
Added an additional test to check on the bg thread creation and closing the CertClient
Also I ran the TestHddsSecureDatanodeInit.testCertificateRotation test 300 times without failure locally, before patch, from 100 runs I was able to see 4 failures.

### CODE REVIEW

File: hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/client/TestDefaultCertificateClient.java [hunk 1]
```diff
@@ -532,4 +544,53 @@ public void testRenewAndStoreKeyAndCertificate() throws Exception {
     // a success renew after auto cleanup new key and cert dir
     dnCertClient.renewAndStoreKeyAndCertificate(true);
   }
+
+  @Test
+  public void testCloseCertificateClient(@TempDir File metaDir)
```
Comment (kerneltime) [inline hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/client/TestDefaultCertificateClient.java:549]: Other than calling close at the end what is the intention for this new test?

Comment (fapifta) [reply hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/client/TestDefaultCertificateClient.java:549]: We have a certificate lifetime monitor, that is scheduled to be run by a scheduled executor service.
The intention here is to check if after we close the client (which should shut down the executorservice), there are no monitoring thread scheduled to be run. As most of the API is private, the test validates the effects that are visible outside.

Once the client is instantiated, there has to be a scheduled thread with the name compName + "-CertificateLifetimeMonitor", then when the client is closed, and the thread is terminated by the shutdown call in the scheduled executor, the test checks that the thread exists after instantiation, and then it checks there is no thread anymore with the name we expect for the monitor thread after close.

Comment (kerneltime) [reply hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/client/TestDefaultCertificateClient.java:549]: Can you please add the same to the test Java doc?

Comment (fapifta) [reply hadoop-hdds/framework/src/test/java/org/apache/hadoop/hdds/security/x509/certificate/client/TestDefaultCertificateClient.java:563]: yes, done.
I also notified two other places where we potentially do not close a client, I have added that as well.

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java [hunk 2]
```diff
@@ -198,14 +198,18 @@ connectionFactory, getScmDBSnapshotUrl(),
...
-                    reconStorage, null, null))) {
-              downloadClient.download(targetFile.toPath()).get();
-            } catch (ExecutionException | InterruptedException e) {
-              LOG.error("Rocks DB checkpoint downloading failed", e);
-              throw new IOException(e);
+            SecurityConfig secConf = new SecurityConfig(configuration);
+            try (ReconCertificateClient certClient =
+                     new ReconCertificateClient(
+                         secConf, reconStorage, null, null)) {
+              try (
```
Comment (Galsza) [inline hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java:205]: nit: I'm fine with the nested try catch block, but is there an advantage to it compared to the try with multiple resources version? Either way is ok to me.

**Comment by Galsza** [comment]
Thanks @fapifta for working on this. The changes overall look good, I've left a minor question.

I've found a couple of places missing the closing of clients:
`HASecurityUtils#initializeSecurity`
`TestContainerServer`
Could you please fix these?

**Comment by fapifta** [comment]
`HASecurityUtils#initializeSecurity` iscalled only from init or bootstrap, after which SCM does not remain running, so even if there is a client created in that phase the certificate should not (in regular circumstances) expire within the grace period, so the one background thread that is created does not remain live, once the process exits within (in regular circumstances) a few seconds. Though we can close the client in this method it does not really matter.

Similarly in `TestContainerServer` the client is created in a method that is annotated with `@BeforeAll` and then it is passed to gRPC server creation calls, where it is not re-created, so again we can close it there, but it does not really matter as we do not have parallel test execution.

I have added the closure to these cases as well, to be consistent with the rest. Will commit it with the fix for the integration test failure which is related.

Thank you for the review!

File: hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java (hunk 2 above)
Comment (fapifta) [reply hadoop-ozone/recon/src/main/java/org/apache/hadoop/ozone/recon/spi/impl/StorageContainerServiceProviderImpl.java:205]: Yes, nice catch, and you are right, there is no reason to keep these two in separate try-with-resources constructs, I just blindly applied the solution to add a try ( before the client creation and the rest, and did not noticed the internal try-with-resources. Fixing this one as well shortly.

**Review by Galsza — APPROVED** [review]
Thank @fapifta for working on this patch. Looking good to me. +1

**Pull request merged**